### PR TITLE
Biorxiv crawler

### PR DESCRIPTION
**Pull request by tekrajchhetri**
This pull request consists of the following:

- code to crawl and download the publications from biorxiv based on specified categories.

- code to crawl and download the publications metadata from biorxiv based on specified categories.

### CODE REVIEW

File: biorxiv/crawl_extract_metadata_csv.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+from concurrent.futures import ThreadPoolExecutor, as_completed
+import sys
+def get_user_agents(filepath="user_agents.json"):
+    with open(filepath, "r") as f:
+        try:
+            data = json.load(f)
+            return data["user_agents"]
+        except json.JSONDecodeError as e:
+            sys.stdout.write(f"Error decoding JSON: {e}")
+            return None
```
Comment (aaronkanzer) [inline biorxiv/crawl_extract_metadata_csv.py:55]: do you want to raise some error here? Or is it intentional that failure is OK by returning `None`?

File: biorxiv/crawl_extract_metadata_csv.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+# Function to extract information from an article
+def extract_article_info(article_link):
+    headers = {"User-Agent": get_random_user_agent()}
+    try:
+        article_response = requests.get(article_link, headers=headers)
+        article_response.raise_for_status()
+    except requests.RequestException as e:
+        sys.stdout.write(f"Failed to retrieve article page: {article_link} due to: {e}")
+        return None
```
Comment (aaronkanzer) [inline biorxiv/crawl_extract_metadata_csv.py:73]: Similar to https://github.com/sensein/crawler_publications/pull/1/files#r1644435435

File: biorxiv/crawl_extract_metadata_csv.py (hunk 1 above)
Comment (tekrajchhetri) [reply biorxiv/crawl_extract_metadata_csv.py:55]: No, I don't want to raise any error. It's intentional

File: biorxiv/crawl_extract_metadata_csv.py (hunk 2 above)
Comment (tekrajchhetri) [reply biorxiv/crawl_extract_metadata_csv.py:73]: It's intentional

File: biorxiv/crawl_extract_metadata_csv.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                        article in articles]
+
+                    for future in as_completed(futures):
+                        result = future.result()
+                        if result:
+                            writer.writerow(result)
+                            sys.stdout.write(f"Extracted: {result[0]}")
+
+                current_page += 1
+                time.sleep(random.uniform(5, 10))  # Random delay between 5 and 10 seconds
```
Comment (aaronkanzer) [inline biorxiv/crawl_extract_metadata_csv.py]: I understand you are using the sleep function so that biorxiv doesn't flag your script -- can this be shorter though? seems like a big bottleneck?

**Comment by aaronkanzer** [comment]
Looks good to me overall -- a couple thoughts:

• You have a lot of hard-coded text strings -- perhaps organize as constants at the top of your script just to simplify things?
• Perhaps a few unit tests could go a long way here to make sure that the output is consistently what you assume and covers any edge cases you can thing of

File: biorxiv/crawl_extract_metadata_csv.py (hunk 3 above)
Comment (tekrajchhetri) [reply biorxiv/crawl_extract_metadata_csv.py]: I was using smaller one but then biorxiv was not responding. But I can test shorter time and update accordingly

**Comment by tekrajchhetri** [comment]
> Looks good to me overall -- a couple thoughts:
> 
> • You have a lot of hard-coded text strings -- perhaps organize as constants at the top of your script just to simplify things? • Perhaps a few unit tests could go a long way here to make sure that the output is consistently what you assume and covers any edge cases you can thing of

@aaronkanzer I have added the unit test as well as organized the hard-coded strings as suggested!

File: biorxiv/crawl_download_pdf.py
Comment (puja-trivedi) [inline biorxiv/crawl_download_pdf.py:1]: you might want to look into Python's `logging` module instead of `sys.stout.write`. It has additional features that might be useful and allows you to specify logging levels.

File: biorxiv/crawl_download_pdf.py [hunk 4]
```diff
@@ -55,7 +60,7 @@ def crawl_and_download_pdf(category, output_folder="downloaded_pdf_files", max_p
     with requests.Session() as session:
         while current_page < max_pages:
```
Comment (puja-trivedi) [inline biorxiv/crawl_download_pdf.py]: should the max page be included (because current_page is set to 1)? i.e `while current_page <= max_pages`

File: biorxiv/crawl_download_pdf.py
Comment (puja-trivedi) [inline biorxiv/crawl_download_pdf.py:1]: you might want to add input validation for the command-line arguments in `if __name__ == "__main__"`

File: biorxiv/crawl_extract_metadata_csv.py
Comment (puja-trivedi) [inline biorxiv/crawl_extract_metadata_csv.py:1]: similar to the other file, you might want to look into the `logging` module

Comment (puja-trivedi) [inline biorxiv/crawl_extract_metadata_csv.py:1]: more stylistic changes:
1. maybe add type hints to function signatures for better code clarity and type checking
2. try using black or another PEP 8 compliant formatter

Comment (tekrajchhetri) [reply biorxiv/crawl_extract_metadata_csv.py:1]: Thanks for the suggestion. I have updated the code + formatted using black.

Comment (tekrajchhetri) [reply biorxiv/crawl_extract_metadata_csv.py:1]: If either of you (@puja-trivedi or @aaronkanzer) could approve the review then I could merge.

Comment (puja-trivedi) [reply biorxiv/crawl_extract_metadata_csv.py:1]: @tekrajchhetri - before I approve the review, did you see my comment about the max_page conditional? Is my understanding incorrect? Why should it just be `<` instead of `<=`

**Review by puja-trivedi — APPROVED** [review]

File: biorxiv/crawl_download_pdf.py (hunk 4 above)
Comment (tekrajchhetri) [reply biorxiv/crawl_download_pdf.py]: @puja-trivedi the max pages grows based on the number of publications submitted to biorxiv and is the number of paginated pages

File: biorxiv/crawl_extract_metadata_csv.py
Comment (tekrajchhetri) [reply biorxiv/crawl_extract_metadata_csv.py:1]: Yes, I have seen your comment and responded to it as well. I've added the validation to max pages available as of now
